File: treeBuilder/src/main/java/pl/michalklecha/sns/treeBuilder/logic/tree/Tree.java
package pl.michalklecha.sns.treeBuilder.logic.tree;

import pl.michalklecha.sns.treeBuilder.model.ItemsWithTids;

public abstract class Tree {
    Node root;

    public Tree(Node node) {
        root = node;
    }

    public Tree(ItemsWithTids rootData) {
        root = new Node(rootData);
    }

    public Node getRoot() {
        return root;
    }

    abstract public Tree getSubtree(String label);

    @Override
    abstract public String toString();
}

File: treeRest/src/main/kotlin/pl/michalklecha/sns/treerest/model/Tree.kt
package pl.michalklecha.sns.treerest.model

import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import pl.michalklecha.sns.treerest.dataLoader.jackson.deserializers.TreeDeserializer

@JsonDeserialize(using = TreeDeserializer::class)
data class Tree(private val root: Node) {
    private val index: MutableMap<String, MutableSet<Int>> = mutableMapOf()

    fun index() {
        index(root)
    }

    private fun index(node: Node) {
        node.items.forEach { addToIndex(it, node.ids) }
        node.children.forEach { index(it) }
    }

    private fun addToIndex(item: String, ids: List<Int>) {
        if (!index.containsKey(item)) {
            this.index[item] = mutableSetOf()
        }
        this.index[item]?.addAll(ids)
    }

    fun getIds(item: String): Set<Int> {
        return index[item]?: mutableSetOf<Int>() as Set<Int>
    }
}